### ISSUE TITLE
Mobile Vaccines

### ISSUE THREAD

**Issue by joshxg** [openmsupply/mobile#3161]
## Description

This issue is an epic for the vaccines feature

## RESOURCES
- Google drive folder: https://drive.google.com/drive/u/1/folders/1ONJjqCd8IFae91CWWrbRSHBiFLO_CqQ2
- Features list: https://docs.google.com/document/d/14qf35a4FQfGi-f7Y9j5F-iGKay_49vEspEu2wlSEsPo/edit
- Requirements: https://docs.google.com/spreadsheets/d/1ao_quJ6mUgFQbxKLjacBTj_Dw72lQlUCi9TirRLwGQc/edit#gid=0
- Misc Links: https://docs.google.com/document/d/1adET2W4SHnOHyLAFTQi4_v2XnOMCa7K9HoLCeHgyKNI/edit
- Terminology: https://docs.google.com/document/d/1Zj7zDKBLFpgnlNfYrKyL974r_lY7XzC7vkLUY_i9Las/edit
- DRAFT USER GUIDE DOCS: https://docs.google.com/document/d/1MD17DWP5BJYVh7KuHsdexHUhSMBB0JWg8TeBAR4_gBY/edit#heading=h.ktu9skw3ggj0

## TASKS

- [x] Bluetooth Thunks
    - ISSUE: #3319 
    - PR: 
      - #3366 
      - #3369 
      - #3357 
      - #3360 
- [x] New Vaccine Page
	- ISSUE: #3346 
	- PR: #3409 
- [x] Sensor Detail Page
	- ISSUE: #3347 
	- PR: #3382 
- [x] New Sensor Flow
	- ISSUE: #3348 
	- PR:
	  - #3361 
	  - #3363 
	  - #3367 
- [x] Edit Sensor Flow
	- ISSUE: #3349 
	- PR: #3379 
- [x] Exporting Data
	- ISSUE: #3350 
	- PR: #3447 

### DOCS

- [ ] User Guide
    - Draft: https://docs.google.com/document/d/1MD17DWP5BJYVh7KuHsdexHUhSMBB0JWg8TeBAR4_gBY/edit#heading=h.ktu9skw3ggj0
    - Published:

**Comment by joshxg** [comment]
Just a starting point for this work.

I think some things I would like to clarify are:


- Do we want to consolidate logic between CCE app / mobile?

Lots of benefits, of course. I am currently trying to evaluate different ways of doing this (in my head) and trying to determine some estimates on how long this would take.

- Do we want this to be the start of a transition to SQLite?

I am not convinced that SQLite is a better choice over realm in general (async v sync, bridge issues etc), however due to debugging issues (Which could be solved by using realm differently) as well as their lack of supporting of hermes I think it is best to switch and this is as good a time as any to start making that transition. It would be good to start thinking of a new db layer which could be used for the vaccine tables, at least, and some thought toward how this would look moving forward

**Comment by mark-prins** [comment]
> Do we want to consolidate logic between CCE app / mobile?

I think yes. There was talk of refinements to be made in CCE, which will have to be replicated in mobile later, if the changes are done after this mobile work. Consolidating would reduce the rework, assuming that mobile and CCE remain separate apps.

> Do we want this to be the start of a transition to SQLite?

Again a yes from me - assuming we do go with SQLite. seems like an opportunity to make a start, if this doesn't add too much extra work.

**Comment by Chris-Petty** [comment]
> * Do we want to consolidate logic between CCE app / mobile?
> 
> * Do we want this to be the start of a transition to SQLite?
> 

Ideally yea on both counts like Mark. Only reservation is estimating the work, when exactly this is needed by and balancing it against other things that may be in your pipeline? (Done and dusted by mid Nov. like many other things atm?) At the end of the meeting the other day, the conclusion I gathered was to do the quicker copy-paste-ish solution you can right?

**Comment by joshxg** [comment]
I've updated the top comment with slightly more detail and some extra stuff I forgot.

**This is a bit longer than I expected - soz**

Lets try and address this first: 

> Do we want to consolidate logic between CCE app / mobile?

I am going to have to be a put pushy on using the module https://github.com/Polidea/react-native-ble-plx over our current ble module https://github.com/openmsupply/mobile/tree/develop/android/app/src/main/java/com/msupplymobile/Ble - not only will this make it easier to implement the programming features but the bluetooth capabilities are enhanced (speed/reliability) as well as the experience (user and developer) is a lot better overall.

This will not take too long as I can use this wrapper from CCE https://github.com/openmsupply/msupply-cold-chain/blob/develop/src/common/services/Bluetooth/BleService.js which has a very similar API to what we use in mobile.

As for other logic (Creating and saving temperature logs from the binary coming from the bluetooth, creating breaches etc). This is the architecture of the CCE app currently:

![image](https://user-images.githubusercontent.com/35858975/95388877-7a120600-094f-11eb-8a36-d83bc6f1ee1f.png)

I have an idea - Currently the logic for these things are within objects which are coupled with the database service of the CCE app, with is SQLite. This was done for simplicity. So this would be something like:

```
const downloadManager = new DownloadManager(databaseService, ..)
const downloadedLogs = await downloadManager.downloadLogs(macAddress)
```

Which would download the logs as well as save them to the database. The `downloadManager` is dependent on a `databaseService`. We could create an adapter: `RealmAdapter` which would mimic the database service API required by the manager objects and map that to some function which uses the realm API [The `databaseService has an API such as `getAll` and `upsert` which are easily mimic'd].

So, in CCE it would be something like:

`const downloadManager = new DownloadManager(databaseService, ..)`

Then, in mobile it would be something like:

`const downloadManager = new DownloadManager(new RealmAdapter(UIDatabase), ...)`

Where the `downloadManager` could be used within `thunks` rather than `sagas` like it is in CCE.

I'm really not sure on how to share this code exactly without just copy pasting. An NPM package doesn't quite feel right!

---------

The above doesn't quite solve everything. It is introducing a new design into mobile, where there are objects which encapsulate some logic, which are NOT realm classes. This pattern leans to a data mapper data access/ORM layer pattern rather than an active record pattern. I would like to ensure that when/if we go to SQLite, I have not started some pattern which will conflict with whatever way we want to go with that.


----------


*Slight tangent*

This also doesn't share the redux/saga code. I am hesistant to share that code (all you need to do at that point is share the components and then it is litterally the same app!) as it seems as if that is too app-specific- let alone mobile using thunks not sagas, so having two async tools might be a little much? Although I am wondering about a way to extract a full 'vertical slice', where you export a reducer and actions (For say, bluetooth) - but this *is* coupled to the database implementation... 🤷

**Comment by mark-prins** [comment]
I had in mind an npm package to share the code. Or a git submodule? Or copy and paste - less ideal, but less work:)
Alternatively, we use something like nx and make a monorepo - which is likely more work than we want to do right now.

Are there options for dependency injection? That would be a nice approach, rather than passing in the database service when instantiating. Not sure about javascript options for DI sorry, would have to explore that. It's just a nice pattern to use.

**Comment by joshxg** [comment]
> I had in mind an npm package to share the code. Or a git submodule? ....

Cool - I'll try and think of the exact scope of what the module will do. Would be good if it did not have dependencies

> Are there options for dependency injection? That would be a nice approach, rather than passing in the database service when instantiating. Not sure about javascript options for DI sorry, would have to explore that. It's just a nice pattern to use

I don't think there's any nice JS DI frameworks if thats what you're asking - just constructor/setting di - I maybe am not clear on what you're asking though!

**Comment by joshxg** [comment]
In regards to SQLite:


A straight replacement of Realm with SQLite is really quite easy and quick job ;-).  I see it potentially taking about a week (so, lets say two? ;-)) - there are minimal changes needed and it is really possibly to just write the schema and replace all (most) of the calls to call to SQLite (Seriously, 95% of the places where database calls are, it's OK to just replace).

There are two main issues: 

1. Where to put the business logic which is currently located in the realm classes. 
2. Slowness

I'm going to put 1 to the side for now *(would love it if anyone reading this gave it a bit of thought and let me know their ideas at some point, though!)*. As for 2: there is an implicit ~40ms of lag for any call to sqlite. More Info: https://docs.google.com/document/d/1_FEkGicLgD3zeADubFUiXw086QOaexotLBclbO_pidM/edit and the current design writes to the database A LOT. Expanding on that:

- Sync writes to the database for every individual record. This means, 100 records * 40ms = 4seconds or 30,000 records (not unlikely) would be about 20 minutes just of writing to the db.
    - Need change up the incoming sync. Writing to the DB per record isn't gonna fly? Need to have each record concat to a single SQL statement to be executed once.
    - Outgoing sync: Might need to test this more. I think it would be *ok*, but still a lot slower than now. Might consider syncing on a per-table ordering rather than a time based ordering though..
- When we edit cells, we write directly to the db (So, for example we currently: 1. edit cell 2. write to db 3. re-render object in interface). This method with SQLite is less ideal and causes at least 40ms of lag on each keystroke - pretty yuck. We can fix this through: 
    - Save after editing - i.e a "SAVE" button.
    - Just queue up writes: 1. input into cell 2. throw a promise to update db and don't wait 3. update ui state

**Comment by Chris-Petty** [comment]
1. Where go classes/business logic?

Write an ORM? Suddenly less small of a job? Hopefully could reuse the existing schema classes in some manner.

2. Slow; each DB call has 40ms lag (until RN is improved (ETA: soon<sup>tm</sup>))

Cell edits: RN aims for >=60hz rendering for that smooth experience. That's about 16ms per frame. I'd guess we wouldn't want keystrokes to have any more than a 50ms render time. 40ms gets us awfully close to that.

My thoughts over the weekend, how hard would it be to implement a DB access class/object that debounced/batched writes and/or async to boot? I think it'd be good to keep the concept of "saves as you type", though there is always fair argument that we lose the ability for a clean "cancel" button to discard changes. This would potentially cover both user input and large writes quite nicely.

**Comment by joshxg** [comment]
> Write an ORM? Suddenly less small of a job? Hopefully could reuse the existing schema classes in some manner.

Writing an ORM would be a pretty big job! There are plenty of made solutions, though. There are two patterns here: Active record (`someTransaction.finalise()`) or data mapper (`TransactionManager.finalise(someTransaction)`. We use a sort of mixture by using `UIDatabase.save()` for example for sync triggers, but active record for actual database changes.

Active record is nice and will allow for probably an easier transition. Data mapper will allow us to work with plain JS objects, rather than class instances - enabling redux debugging and simpler react-rendering logic (i.e no need for the `dataState` `refreshRow` shenanigans in the data table to be re-render rows)

> how hard would it be to implement a DB access class/object that debounced/batched writes and/or async to boot? 

Super easy! - but not sure if this fixes the problem? (Might be thinking about it wrong). Currently we `edit -> write to db -> re-render new db object` If we throttle and buffer, then we also need to solve the problem of what is re-rendered. I do like this idea - a slightly different approach could be a HOC which buffers inputs and debounces at the UI level.



-----

For reference - currently we use this case reducer to re-render rows in the data table as we keep class instances of realm objects which make re-rendering immutably a lot harder:

```
export const refreshRow = (state, action) => {
  const { payload } = action;
  const { rowKey } = payload;
  const { dataState } = state;

  // Change object reference of row in `dataState` to trigger rerender of that row.
  // Realm object reference in `data` can't be affected in any tidy manner.
  const newDataState = new Map(dataState);
  const nextRowState = newDataState.get(rowKey);
  newDataState.set(rowKey, { ...nextRowState });

  return { ...state, dataState: newDataState };
};
```

Essentially we keep a `Map` which is in sync with the actual data and set a new object in that `Map` to trigger a re-render on a row. Technically this is because realm class instances are instantiated by `Realm`. But, I am not sure that it's specifically a Realm issue? If they were not realm classes we could re-instantiate an instance instead: `const newTransaction = new Transaction(oldTransaction)` and have the constructor use the seed values.

It really is just nicer to use plain JS objects, though: `const newTransaction = {...oldTransaction }`

Hope this gives some helpful insight..!

**Comment by andreievg** [comment]
My 2 cents:

### Log Interval

A big question for me is the restricted bandwidth assumption that we've previously made. Calculation extract:

1 sensor log:

ID 32bytes
LocationUUID 32bytes
SensorUUID 32bytes
Timestamp 64bytes
BreachUUID 32bytes
StoreUUID 32bytes

`224 bytes` (which is actually quite a bit of stored data).

1 per minutes = 224 * 1440 = `323kb per day` (pretty small imo)
Likely a site will not have data for a month = `~10mb per month`

I think thats ok

Our second worry was that server might have too many records. We could make an issue to start archiving/aggregating at 4D end, if number of records is deemed excessive.

#### UI Wise

* Export
  * I think there is enough real estate on main VM page for a button 
* Selecting date range (already exists)
 * Cumulative can just be displayed in the fridge header in VM (where existing info is displayed, can make this a double row header)
* Clearing breaches
  * Can discuss this, but since we show individual breaches on the graph, don't see a need to clear them. If we do need to  show and clear, them, then have another bit of info in the header 'New Breaches', with number of breaches, can press and clear.
* Configuring time delay? Can this be an option when sensor is linked to a fridge ? I cant's quite see how this is done in current VM (https://docs.msupply.foundation/en:mobile:user_guide:cold_chain_equipment), do we have a full set of screenshots or a screencast of functionality ?

Multi sensor/multi fridge ? Is this possible in current VM ?

##### Logic

I like the idea of sharing this logic between apps. I am not sure why it would be hundreds of hours of work, would be great to know why this approach wouldn't' work:

* move logic into github package.
* both apps have different' entry point for the logic, but still use the same generator methods and sql methods etc..

A couple of difficulties that i see in using SQLite in parallel to Realm:
* Links between entities in seperate DBs
  * ID's where needed instead of object relations
  * Realm data class getters to use sqlite singleton
* Sync
  * Incoming, createOrUpdateRecord to insert into appropriate DB, and use ID rather the relation (yep, this will be slow for mass number of records, for now it should be sufficient. otherwise we make mechanism to add to 'to be saved' queue, and run mass save at the end of sync operation) 
  * Outgoing, another sync queue, in sqlite db, add triggers to table to populate it

* Passive Bluetooth Sync
  * I like the idea of using logic from Cold Chain app for this.

#### Configuration of breach 

Configurations for breaches -> desktop 

#### Further SQLite thoughts:

Btw I think we should move non Vaccine mobile talk to another issue. 

> This method with SQLite is less ideal and causes at least 40ms of lag on each keystroke - pretty yuck.

I was wondering, is it that bad ? Isn't the responsiveness rule to keep it under 0.1s. And would you need to type at the speed of 25 characters per second to notice lag ?

Also, I am not sure that saving to DB/filesystem per key stroke is intended use of hardware. (and I am pretty sure realm doesn't save to hard driver on every write() statement). Although when we do go to SQLite, I think MVP wise, it should be ok. Ideally validate onChange, save to db on loose focus ?

For sync, we should definitely do a mass update, one statement per sync bunch.

**Comment by andreievg** [comment]
I didn't quite comment on the thunk/saga difference, was wondering why not use them both if they are used for functionality in certain their own domains ?

**Comment by joshxg** [comment]
Have gotten design hifi proto type back from the designer https://marvelapp.com/prototype/55aea8f/screen/75648032

Will be breaking this down unto individual issues
